File: ios/Runner/Handlers/MethodHandler.swift
//
//  MethodHandler.swift
//  Runner
//
//  Created by GFWFighter on 10/23/23.
//

import Flutter
import Combine
import RaynCore

public class MethodHandler: NSObject, FlutterPlugin {
    
    private var cancelBag: Set<AnyCancellable> = []
    
    public static let name = "\(Bundle.main.serviceIdentifier)/method"
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: Self.name, binaryMessenger: registrar.messenger())
        let instance = MethodHandler()
        registrar.addMethodCallDelegate(instance, channel: channel)
        instance.channel = channel
    }
    
    private var channel: FlutterMethodChannel?
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        @Sendable func mainResult(_ res: Any?) async -> Void {
            await MainActor.run {
                result(res)
            }
        }
        
        switch call.method {
        // Returns the PEM certificate the Dart client pins for the secure gRPC
        // modes. This used to `result("")` -- a stub that looked implemented,
        // which is part of why the secure modes were believed to work.
        //
        // Delivered over the method channel deliberately: that is in-process and
        // therefore trustworthy, whereas fetching it over the very connection it
        // is meant to authenticate would authenticate nothing. Only valid after
        // `setup` has run with a secure mode; before that the core has no
        // certificate and this correctly returns nil.
        case "get_grpc_server_public_key":
            if let pem = MobileGetServerPublicKey(), !pem.isEmpty {
                result(FlutterStandardTypedData(bytes: pem))
            } else {
                result(nil)
            }
        // Client certificates are not used -- the client authenticates with the
        // setup secret. Kept so an older Dart build calling it gets a clear
        // answer rather than MissingPluginException.
        case "add_grpc_client_public_key":
            result(FlutterError(code: "UNSUPPORTED",
                                message: "client certificates are not used; the client authenticates with the setup secret",
                                details: nil))
        // The credential Dart attaches to every RPC. Creating on first call is
        // correct here and only here: this runs in the app, whereas the extension
        // only ever peeks, so a background start cannot race the app into
        // generating a second secret and leave the two cores disagreeing.
        case "get_grpc_secret":
            if let secret = GrpcSecret.getOrCreate() {
                result(secret)
            } else {
                result(FlutterError(code: "GRPC_SECRET",
                                    message: "keychain unavailable",
                                    details: nil))
            }
        case "setup":
                Task {
                    guard
                        let args = call.arguments as? [String: Any?],
                        let baseDir = args["baseDir"] as? String,
                        let workingDir = args["workingDir"] as? String,
                        let tempDir = args["tempDir"] as? String,
                        let mode = args["mode"] as? Int,
                        let grpcPort = args["grpcPort"] as? Int
                    else {
                        result(FlutterError(code: "INVALID_ARGS", message: nil, details: nil))
                        return
                    }
                    // Dart does not send `debug` on every path; absent means off.
                    let debug = (args["debug"] as? NSNumber)?.boolValue ?? false
                    VPNConfig.shared.baseDir=baseDir
                    VPNConfig.shared.workingDir=workingDir
                    VPNConfig.shared.tempDir=tempDir
                    var error: NSError?
                    let opts = MobileSetupOptions()
                    opts.basePath = baseDir
                    opts.workingDir = workingDir
                    opts.tempDir = tempDir
                    opts.listen = "127.0.0.1:\(grpcPort)"
                    // The credential the core requires on every RPC. Was hardcoded
                    // empty, and the core read it into nothing -- plumbed the whole
                    // way here and used by no one.
                    //
                    // Owned natively rather than passed from Dart, because the
                    // packet-tunnel extension needs the SAME value and an
                    // on-demand start has no app process to hand it one. Dart
                    // fetches it back over `get_grpc_secret` for its call metadata.
                    opts.secret = GrpcSecret.getOrCreate() ?? ""
                    // Both were hardcoded, ignoring the arguments parsed above.
                    // `mode` mattered twice over: the app process is the
                    // FOREGROUND core (Dart sends 3 = GRPC_NORMAL_INSECURE,
                    // rayn_core_service.dart:99) and pinning it to 4 meant
                    //   * it set up as a background core while Dart later closed
                    //     mode 3 (rayn_core_service.dart:607) — the wrong server;
                    //   * hcore redirects stderr to data/stderr<mode>.log
                    //     (v2/hcore/grpc_server.go:67), so this process and the
                    //     packet-tunnel extension — which is legitimately mode 4 —
                    //     interleaved into one file.
                    // Android has always forwarded both (MethodHandler.kt:94).
                    opts.debug = debug
                    opts.mode = mode
                    opts.fixAndroidStack = false
                    MobileSetup(opts,
                        nil,
                        &error
                    )
                    
                    if let error {
                        result(FlutterError(code: String(error.code), message: error.localizedDescription, details: nil))
                        return
                    }
                    do {
                        try await VPNManager.shared.setup()
                    } catch {
                        result(FlutterError(code: "SETUP", message: error.localizedDescription, details: nil))
                        return
                    }
                    result(true)
                }
        case "start":
            Task {
                guard
                    let args = call.arguments as? [String:Any?],
                    let path = args["path"] as? String,
                    let name = args["name"] as? String,
                    let grpcPort=args["grpcPort"] as? Int
                else {
                    await mainResult(FlutterError(code: "INVALID_ARGS", message: nil, details: nil))
                    return
                }
                VPNConfig.shared.activeConfigPath = path
                VPNConfig.shared.activeProfileName = name
                VPNConfig.shared.grpcServiceModePort=grpcPort
                
                var error: NSError?
                //let configstr=MobileBuildConfig(path,&error) as String
                if let error {
                    await mainResult(FlutterError(code: String(error.code), message: error.description, details: nil))
                    return
                }
                do {
                    try await VPNManager.shared.setup()
                    try await VPNManager.shared.connect(with: path, grpcServiceModePort: grpcPort, disableMemoryLimit: VPNConfig.shared.disableMemoryLimit)
                } catch {
                    await mainResult(FlutterError(code: "SETUP_CONNECTION", message: error.localizedDescription, details: nil))
                    return
                }
                await mainResult(true)
            }
//        case "restart":
//            Task { [unowned self] in
//                guard
//                    let args = call.arguments as? [String:Any?],
//                    let path = args["path"] as? String,
//                    let name = args["name"] as? String,
//                    let grpcPort=args["grpcPort"] as? Int
//                else {
//                    await mainResult(FlutterError(code: "INVALID_ARGS", message: nil, details: nil))
//                    return
//                }
//                VPNConfig.shared.activeConfigPath = path
//                VPNConfig.shared.activeProfileName = name
//                VPNConfig.shared.grpcServiceModePort=grpcPort
//                
//                
//                VPNManager.shared.disconnect()
//                await waitForStop().value
//                var error: NSError?
//                do {
//                    try await VPNManager.shared.setup()
//                    try await VPNManager.shared.connect(with: path, disableMemoryLimit: VPNConfig.shared.disableMemoryLimit)
//                } catch {
//                    await mainResult(FlutterError(code: "SETUP_CONNECTION", message: error.localizedDescription, details: nil))
//                    return
//                }
//                await mainResult(true)
//            }
        case "stop":
            VPNManager.shared.disconnect()
            result(true)
        case "reset":
            VPNManager.shared.reset()
            result(true)
        // Removed with this comment as their headstone, so nobody re-adds them:
        //
        //   select_outbound, url_test  -- called LibboxNewStandaloneCommandClient()
        //     against a command server this extension has not started since the
        //     gRPC switch (see the commented-out commandServer in
        //     ExtensionProvider). Dart reaches both over gRPC instead
        //     (rayn_core_service.dart:358 -> bgClient.selectOutbound), which is
        //     also why the sing-box 1.14 selection-notify fix applies to iOS
        //     without anything extra.
        //   generate_config, parse_config -- bodies were commented out.
        //     generate_config additionally never called result(), so invoking it
        //     would have hung the Dart future forever.
        //   generate_warp_config -- WARP was removed from this fork.
        //   change_hiddify_options -- wrote VPNConfig.configOptions, which
        //     nothing reads; the config now arrives from Dart already built.
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func waitForStop() -> Future<Void, Never> {
        return Future { promise in
            var cancellable: AnyCancellable? = nil
            cancellable = VPNManager.shared.$state
                .filter { $0 == .disconnected }
                .first()
                .delay(for: 0.5, scheduler: RunLoop.current)
                .sink(receiveValue: { _ in
                    promise(.success(()))
                    cancellable?.cancel()
                })
        }
    }
}

File: android/app/src/main/kotlin/com/raynlabs/app/MethodHandler.kt
package com.raynlabs.app

import android.util.Log
import com.raynlabs.app.bg.BoxService
//import com.raynlabs.app.bg.BoxService.Companion.workingDir
import com.raynlabs.app.constant.Status
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

import com.raynlabs.core.libbox.Libbox
import com.raynlabs.core.mobile.Mobile
import com.raynlabs.core.mobile.SetupOptions
import com.raynlabs.app.bg.Bugs
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.io.File

class MethodHandler(private val scope: CoroutineScope) : FlutterPlugin,
    MethodChannel.MethodCallHandler {
    private var channel: MethodChannel? = null

    companion object {
        const val TAG = "A/MethodHandler"
        const val channelName = "com.raynlabs.app/method"

        enum class Trigger(val method: String) {
            Setup("setup"),
            Start("start"),
            Stop("stop"),
            Restart("restart"),
            AddGrpcClientPublicKey("add_grpc_client_public_key"),
            GetGrpcServerPublicKey("get_grpc_server_public_key"),
            GetGrpcSecret("get_grpc_secret"),
            RotateGrpcBgSecret("rotate_grpc_bg_secret"),
            GetGrpcBgSecret("get_grpc_bg_secret"),

        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(
            flutterPluginBinding.binaryMessenger,
            channelName,
        )
        channel!!.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            // Client certificates are not used — the client authenticates with the
            // setup secret. This previously stored the key in Settings and returned
            // success without ever reaching the core, so it read as implemented
            // while doing nothing. Answer explicitly rather than silently.
            Trigger.AddGrpcClientPublicKey.method -> {
                result.error(
                    "UNSUPPORTED",
                    "client certificates are not used; the client authenticates with the setup secret",
                    null,
                )
            }

            Trigger.GetGrpcServerPublicKey.method -> {
                GlobalScope.launch {
                    result.runCatching {
                        result.success(Mobile.getServerPublicKey())
                    }
                }
            }

            // The credential Dart attaches to every RPC. Generating on first read
            // is safe here because both readers — this process and the VPN service
            // — go through the same SharedPreferences, so whichever asks first
            // wins and the other sees the stored value.
            Trigger.GetGrpcSecret.method -> {
                result.success(Settings.grpcSecret)
            }

            // Minted fresh on every connect, and returned so Dart can attach it to
            // the background channel it is about to use. Dart calls this BEFORE
            // `start`, because the service reads the stored value once when it sets
            // the core up — rotating afterwards would leave the two disagreeing and
            // fail every background call.
            //
            // Rotation is what bounds the exposure of a secret that crosses a
            // plaintext socket: a process that squats the port and captures one
            // holds a value that is dead by the next connect.
            Trigger.RotateGrpcBgSecret.method -> {
                result.success(Settings.rotateGrpcBgSecret())
            }

            // READ, never mint. The VPN service outlives the Flutter engine, so a
            // Dart side that was recreated -- the first-install VPN consent dialog
            // recreates the activity, and process death does the same -- has to be
            // able to adopt the secret the running core was started with. Minting
            // here instead would hand the app a value the core has never seen and
            // wedge every background call.
            Trigger.GetGrpcBgSecret.method -> {
                result.success(Settings.grpcBgSecret)
            }

            Trigger.Setup.method -> {
                GlobalScope.launch {
                    result.runCatching {
                        val args = call.arguments as Map<*, *>
                        Settings.baseDir = args["baseDir"] as String
                        Settings.workingDir = args["workingDir"] as String
                        Settings.tempDir = args["tempDir"] as String
                        Settings.debugMode = args["debug"] as Boolean? ?: false
                        val mode = args["mode"] as Int
                        val grpcPort = args["grpcPort"] as Int
                        Log.d("debugmode","${Settings.debugMode}")
                        runCatching {
                            Mobile.setup(
                                SetupOptions().also {
                                    it.basePath = Settings.baseDir
                                    it.workingDir = Settings.workingDir
                                    it.tempDir = Settings.tempDir
                                    it.fixAndroidStack = Bugs.fixAndroidStack
                                    it.mode=mode.toLong()
                                    it.listen= "127.0.0.1:" + grpcPort
                                    // The credential the core requires on every RPC.
                                    // Owned here rather than passed from Dart: the
                                    // VPN service needs the SAME value and can be
                                    // started by the system with no Flutter engine
                                    // alive. Dart reads it back over get_grpc_secret.
                                    it.secret = Settings.grpcSecret
                                    it.debug = Settings.debugMode
                                },null)

//                            Libbox.setup(Settings.baseDir, Settings.workingDir, Settings.tempDir, false)
                            Libbox.redirectStderr(File(Settings.workingDir, "stderr2.log").path)

                            success("")
                        }.onFailure {
                            error(it)
                        }

                    }
                }
            }


            Trigger.Start.method -> {
                scope.launch {
                    result.runCatching {
                        val args = call.arguments as Map<*, *>
                        Settings.activeConfigPath = args["path"] as String? ?: ""
                        Settings.activeProfileName = args["name"] as String? ?: ""
                        Settings.debugMode = args["debug"] as Boolean? ?: false
                        Settings.grpcServiceModePort = args["grpcPort"] as Int

                        val mainActivity = MainActivity.instance
//                        val started = mainActivity.serviceStatus.value == Status.Started
//                        if (started) {
//                            Log.w(TAG, "service is already running")
//                            return@launch success(true)
//                        }
                        Settings.startCoreAfterStartingService = false

                        mainActivity.startService()
                        success(true)
                    }
                }
            }

            Trigger.Stop.method -> {
                scope.launch {
                    result.runCatching {
                        val mainActivity = MainActivity.instance
                        val started = mainActivity.serviceStatus.value == Status.Started
                        if (!started) {
                            Log.w(TAG, "service is not running")
                            //    return@launch success(true)
                        }
                        BoxService.stop()
                        success(true)
                    }
                }
            }

//            Trigger.Restart.method -> {
//                scope.launch(Dispatchers.IO) {
//                    result.runCatching {
//                        val args = call.arguments as Map<*, *>
//                        Settings.activeConfigPath = args["path"] as String? ?: ""
//                        Settings.activeProfileName = args["name"] as String? ?: ""
//                        val mainActivity = MainActivity.instance
//                        val started = mainActivity.serviceStatus.value == Status.Started
//                        if (!started) return@launch success(true)
//                        val restart = Settings.rebuildServiceMode()
//                        if (restart) {
//                            mainActivity.reconnect()
//                            BoxService.stop()
//                            delay(1000L)
//                            mainActivity.startService()
//                            return@launch success(true)
//                        }
//                        runCatching {
//                            Libbox.newStandaloneCommandClient().serviceReload()
//                            success(true)
//                        }.onFailure {
//                            error(it)
//                        }
//                    }
//                }
//            }

            else -> result.notImplemented()
        }
    }
}